### ISSUE TITLE
Update how Groups and Users are chosen in the metabox

### ISSUE THREAD

**Issue by stevejburge** [publishpress/PublishPress-Permissions#1196]
In the Permissions metabox, we have an old search system.

Can we update this with a modern system, perhaps Select2?
![Screenshot 2024-10-14 at 2 33 39 PM](https://github.com/user-attachments/assets/ee1278cb-6110-45d5-a6ba-9128a285b8b4)
![Screenshot 2024-10-14 at 2 32 59 PM](https://github.com/user-attachments/assets/abf282c9-0e55-4ddd-8668-19ecac003c59)

**Comment by agapetry** [comment]
Any new search solution you settle on should avoid pre-querying all users, as that would cause performance issues on some sites.

**Comment by stevejburge** [comment]
The Select box is good, but some suggested changes:

- The extra "Reading permissions have not been added or blocked for any Custom Group. and "Reading permissions have not been added or blocked for any User. is confusing. We can remove that and 
- The search is only for usernames now. @agapetry - thoughts on that? Allow searching by "Name" also?

![Image](https://github.com/user-attachments/assets/de0a3834-a11b-42c3-80bf-84dd5b3e7b40)

**Comment by agapetry** [comment]
@stevejburge I agree that it should also support searching by display_name or user_email.

**Comment by stevejburge** [comment]
OK, yes, you're correct. I see what you mean, @richaferry. 

I think this is a visual issue. Take this user as an example:

![Image](https://github.com/user-attachments/assets/5e456111-a64c-4a64-b804-fae4f7ae85f8)

Type in "Burge" and this is the result. It's correct, but feels incomplete. Maybe the dropdown shows this way? "First Name Last Name (d

![Image](https://github.com/user-attachments/assets/f17e8968-859f-4fdb-9715-3764e1607c20)

**Comment by richaferry** [comment]
Fixed on this commit [fix formatting name ajax search user](https://github.com/publishpress/PublishPress-Permissions/pull/1506/commits/2a4ca1e7bc9edb46a8dd776e95c6b36de8df89ae)

**Comment by agapetry** [comment]
@richaferry @stevejburge Can this be styled with a search or "dropdown" icon on the right? I think some may find the unadorned blank box puzzling at first glance.

![Image](https://github.com/user-attachments/assets/fb3d1f75-5eff-4c45-ac60-f78864c7e314)

![Image](https://github.com/user-attachments/assets/f6ff9e76-6e20-4235-a5f1-38353bdd53c6)

**Comment by agapetry** [comment]
Yes, I think that would help clarify it. Maybe it could be suppressed if some items are already saved, to save space?

**Comment by agapetry** [comment]
I just noticed that clicking "Clear all" causes "Search for an item" to be displayed:

![Image](https://github.com/user-attachments/assets/1f169d2e-205b-42fc-b6a0-75dc01bc5492)

We either display that from the start when no items are stored, or stop displaying it at all.

**Comment by agapetry** [comment]
Another quirk is that the "create group" link is only displayed if at least one group is already stored:

![Image](https://github.com/user-attachments/assets/6ef0cc48-d721-4893-9816-4380c901fb54)

**Comment by agapetry** [comment]
@richaferry More importantly, there are some errant captions displayed when a page (or other hierarchical post type) is being edited:

![Image](https://github.com/user-attachments/assets/daf9acb6-7af1-4ef5-8f04-41735280465e)

The "This Page" and "Sub-Pages" captions are left over from the previous solution.

**Comment by agapetry** [comment]
@richaferry 

I was expecting the stored items to show **Display Name (username)** as the search results, but they seem to be showing **Display Name (Display Name)**:

![Image](https://github.com/user-attachments/assets/d9281f51-a21b-4821-9928-bb25b8294884)

**Comment by richaferry** [comment]
Did you trying 4.1.4-beta.3 @agapetry ? for display caption in page. because i've remove that code.

**Comment by richaferry** [comment]
For name im use user_login for that. i'll check again. Thanks for your advice @agapetry

**Comment by richaferry** [comment]
How to replicate format name like that @agapetry ? because in my side already use user_login field. 

<img width="536" alt="Image" src="https://github.com/user-attachments/assets/b1fefe80-4fa8-4428-bf26-133d92852f66" />

**Comment by agapetry** [comment]
@richaferry I'm working with the current contents of the [PublishPress-Permissions release-4.1.4 branch](https://github.com/publishpress/PublishPress-Permissions/tree/release-4.1.4).

**Comment by agapetry** [comment]
> How to replicate format name like that [@agapetry](https://github.com/agapetry) ? because in my side already use user_login field.
> 

@richaferry It's correct in the search results, but look at the stored items display:

![Image](https://github.com/user-attachments/assets/2fd0a921-1b70-4994-9b33-d65ac1e06c09)

**Comment by agapetry** [comment]
@stevejburge How about aiming for both simplicity and clarity by way of an "info" icon in that unused space to the right of the selector?

![Image](https://github.com/user-attachments/assets/d2b48938-0bad-44bf-8508-34b038993a7b)

The popup could say "No custom Permission Groups have been stored. [link]Create a group[/link].

Only slightly less simple than an unexplained "Searching... No results," but a lot clearer.

**Comment by stevejburge** [comment]
@agapetry In that vein, we can hide the "Groups" tab if none have been created

**Comment by richaferry** [comment]
I can't replicate your issue @agapetry . in my side items are stored same with search. 
<img width="553" alt="Image" src="https://github.com/user-attachments/assets/c44f3ade-63b0-438b-b43e-de52472b4ab1" />

**Comment by richaferry** [comment]
<img width="557" alt="Image" src="https://github.com/user-attachments/assets/eaa4cd94-8a76-43e4-9c4b-282b8167d045" />

**Comment by agapetry** [comment]
@richaferry Regarding "This Page" and "Sub-Pages," there is actually a more fundamental gap in your solution. For hierarchical post types, two different settings need to be handled:

* customized Permissions for the page itself
* customized Permissions for its sub-pages
 
I'm not seeing any way to view and assign sub-page Permissions. You can see this on the Roles tab, and in version 4.1.3's Groups and Users tabs:

![Image](https://github.com/user-attachments/assets/64203d61-cbf6-46dc-aa2e-5519a5ee6988)

If this is not corrected, updating the Page with new Permissions would (at best) leave existing sub-page permissions hidden and uneditable.  Or at worst (depending on your implementation), it might clear or apply sub-Page Permissions without notice.

**Comment by richaferry** [comment]
Ok i got your point @agapetry . 
What are you suggestion for that case? Is it ok if i just put code back?

<img width="898" alt="Image" src="https://github.com/user-attachments/assets/f7dbd52b-7900-4586-9448-2ab15f6ccca5" />

**Comment by agapetry** [comment]
Yes, you could solve it that way. In fact, you need to bring the dropdowns back for non-hierarchical post types also, for the modification type selection (Enabled / Blocked etc.)

So your Select2 box will only display new selections, not currently stored items (making my reported Display Name (Display Name) issue a moot point).

**Comment by richaferry** [comment]
Wait,
So input search with select2 just show new search only @agapetry ?
Is it ok with that @stevejburge ?

**Comment by stevejburge** [comment]
@richaferry Yes, please. That last screenshot makes sense

Let's hide the "Groups" tab if not groups have been created.

**Comment by richaferry** [comment]
For this requirement 
`Let's hide the "Groups" tab if not groups have been created.`
can we adjust later @stevejburge ? Because in my perspection this mechanism require alot of resource. every time page has refresh, and then will check has group list or not. and current condition to get list group using ajax. if we are moving from ajax into page request it would be make reload longer. And if we are checking using ajax and then it would be give glitch for user experience. and i think it's not good.

**Comment by stevejburge** [comment]
@richaferry Yes, that's fine. Can you start a new issue for that feature, please?

**Comment by agapetry** [comment]
@richaferry @stevejburge  As it stands now, the new solution is technically workable, but in reality a no-go. Here's why:

A page already has one User blocked from reading Sub-Pages. Now I go back in to edit it and I want to block another user from Sub-Pages.

I successfully search and select the user, but now I don't know how to specify that this selection is for blockage of sub-Pages. In fact, I can't. All I can do is update with the selection and let it be treated as "Enabled" for "This Page."

![Image](https://github.com/user-attachments/assets/1eea8854-0a30-4ec8-adb1-52801288f5a2)

This will not be evident unless I reload the Page:

![Image](https://github.com/user-attachments/assets/9d9bfdf2-9b6e-4869-95c3-b0669bc831e5)

After that reload, I can edit the Permission to the desired setting. This would obviously be a step in the wrong direction. I'm sorry, but we didn't think through what a Select2 selection would mean in relation to these available Permissions adjustments. In hindsight, I think the only way it's useful is if we implement separate Select2 controls on the Groups and Users tabs:

- Enabled for This Page
- Blocked for This Page
- Enabled for sub-Pages
- Blocked for sub-Pages

That doesn't seem like it would feel simpler or cleaner, so maybe just use your new search results JS to create a new selection dropdown row as the previous solution did, unless you see a better solution.

**Comment by agapetry** [comment]
Another factor to consider: If you were to implement a Select2 solution to show the current and newly selected users / groups, we would lose the ability to conveniently jump into the Edit Permissions screen for any of those Groups or Users. Each of the user / group captions for current Permissions is a link that would not exist in the Select2 box.

**Comment by stevejburge** [comment]
@agapetry I think we can figure out a solution here. The Select2 is a big usability improvement.

Maybe we can add a button to the search? Click the button to add the user to the right side of the screen?

**Comment by stevejburge** [comment]
The previous interface was hard to use. We need a more modern alternative to this:

![Image](https://github.com/user-attachments/assets/61155530-d42b-42ed-bc77-2dab85534d8f)

**Comment by agapetry** [comment]
That's fine, but it sounds like you are now suggesting to use Richa's new search results solution while dropping Select2.

**Comment by agapetry** [comment]
> Click the button to add the user to the right side of the screen?

We already have that code available to re-implement.

**Comment by stevejburge** [comment]
@agapetry I'm never locked into a specific solution, only looking for improvement.

So would this work?

![Image](https://github.com/user-attachments/assets/df23407e-fb92-4a7c-975c-821fda10e8ac)

**Comment by stevejburge** [comment]
@agapetry do you mind taking over the final step here, please? You know the Permissions codebase better than anyone and I think you're clear on what we've agreed here.

**Comment by agapetry** [comment]
@stevejburge Okay, I'll reinstate my "add user to right side" code and see about making some simple changes to the Select2 configuration to mesh with that.

**Comment by stevejburge** [comment]
Awesome, thanks Kevin
